### PR TITLE
JIT: ensure pred lists stay in increasing bbNum order

### LINKED ISSUE #8720: Predecessor List does not always maintain sorted order.
This repros more deterministically with JitStress = 2; however, it is not necessary. The original assert came from JIT/Directed/cmov/Bool_Xor_Op_cs_do.exe.

To Repro:

Add the following assert to the end of Compiler::fgAddRefPred.

```
#ifdef DEBUG

    if (JitConfig.JitSlowDebugChecksEnabled() != 0)
    {
        // Pedantically make sure bbPreds still in sorted order.

        unsigned prevNum = 0;

        for (flowList* pred = block->bbPreds; pred != nullptr; pred = pred->flNext)
        {
            assert(pred->flBlock->bbNum > prevNum);
            
            prevNum = pred->flBlock->bbNum;
        }
    }

#endif //DEBUG
```

category:correctness
theme:flowgraph
skill-level:intermediate
cost:medium

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Whenever blocks are renumbered or a block is swapped into an existing
pred list entry, ensure the pred list remains properly ordered.

Closes #8720.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@dotnet/jit-contrib PTAL

x64 windows SPMI reports 15 diffs, all minor and in relatively big methods. At first glance these all appear to be block ordering changes going into LSRA, that lead to slightly different allocation decisions.

TP via pin & crossgen of SPC shows just a tiny increase (1870 -> 1871).

Interesting diffs are in `fgRenumberBlocks` and `fgReplacePred`, as well as the new methods on basic blocks.

We should arguably give `flNext` the same sort of protection that this change gives to `flBlock`  (now renamed); there only seems to be a handful of places that try to modify the pred list.

**Comment by AndyAyersMS** [comment]
Failures are caused by code in `fgCompactBlocks` that changes the `bbNum` of a block. Working on a fix.

**Comment by sandreenko** [comment]
It will probably close #10354 also.

**Comment by AndyAyersMS** [comment]
> It will probably close #10354 also.

Thanks.

I couldn't make the test from that issue fail under jit stress with current master. But it seems likely to be fixed, so I'll update the test to run under jit stress going forward.

**Review by sandreenko — APPROVED** [review]
LGTM, I am a bit worried about TP, could you measure it on some nasty tests like https://github.com/dotnet/runtime/tree/master/src/tests/JIT/Methodical/Arrays and run all stress jobs here to see if any timeouts pop-up?

File: src/coreclr/jit/block.cpp [hunk 1]
```diff
@@ -206,6 +208,126 @@ flowList* Compiler::BlockPredsWithEH(BasicBlock* blk)
...
+// Returns:
+//    false if pred list is not in increasing bbNum order.
+//
+bool BasicBlock::checkPredListOrder()
+{
+    unsigned lastBBNum = 0;
+    for (flowList* pred = bbPreds; pred != nullptr; pred = pred->flNext)
+    {
+        const unsigned bbNum = pred->getBlock()->bbNum;
+        if (bbNum <= lastBBNum)
```
[the remark this answers is not in the document]
Comment (AndyAyersMS) [reply src/coreclr/jit/block.cpp:223]: There is such an assert, just below.

File: src/coreclr/jit/block.cpp [hunk 2]
```diff
@@ -206,6 +208,126 @@ flowList* Compiler::BlockPredsWithEH(BasicBlock* blk)
...
+
+    jitstd::sort(sortVector->begin(), sortVector->end(), flowListBBNumCmp());
+
+    // Rethread the list.
+    //
+    flowList* last = nullptr;
+
+    for (flowList* current : *sortVector)
+    {
+        if (last == nullptr)
```
[the remark this answers is not in the document]
Comment (AndyAyersMS) [reply src/coreclr/jit/block.cpp:312]: We catch dups on return to `ensurePredListOrder` by verifying the list is now properly ordered.

Are you suggesting we try and automatically repair lists with duplicate entries?

**Comment by AndyAyersMS** [comment]
Arm failure looks like it might be related --  `System.Text.Tests.TranscodingStreamTests.Read_Span` does not seem to be a known troublemaker.
```
    System.Text.Tests.TranscodingStreamTests.Read_Span(bufferLength: 1) [FAIL]
      Assert.Equal() Failure
                                       â†“ (pos 4401399)
      Expected: Â·Â·Â·vwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdÂ·Â·Â·
      Actual:   Â·Â·Â·vwxyzabcdefghijklmnoPqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdÂ·Â·Â·
                                       â†‘ (pos 4401399)
```
So, will investigate.

File: src/coreclr/jit/block.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (sandreenko) [reply src/coreclr/jit/block.cpp:312]: now when I am thinking about a valid situation where we get a duplicate I can't find any so I agree that we should not try to repair it and just assert.

**Comment by AndyAyersMS** [comment]
> I am a bit worried about TP, could you measure it on some nasty tests

Any test in particular?

**Comment by sandreenko** [comment]
> Any test in particular?

any from huge https://github.com/dotnet/runtime/tree/master/src/tests/JIT/Methodical/Arrays/huge

**Comment by AndyAyersMS** [comment]
I don't see any impact either in checked or release on a handful of those tests.  I think we have other tests that stress the flowgraph a bit more; let me see if I can dig some up.

Pred lists tend to be short and checking for out of order is relatively quick compared to re-ordering, so I think it unlikely that there will be a big impact as we know for the most part pred lists are in order even without this fix. But I'll keep looking.

**Comment by sandreenko** [comment]
Maybe try this with JitStress2 https://github.com/dotnet/runtime/tree/master/src/tests/JIT/Regression/JitBlue/DevDiv_255294

**Comment by AndyAyersMS** [comment]
> JIT/Regression/JitBlue/DevDiv_255294

No impact with or without stress.

**Comment by AndyAyersMS** [comment]
> Arm failure looks like it might be related -- `System.Text.Tests.TranscodingStreamTests.Read_Span` does not seem to be a known troublemaker.
> ...
> So, will investigate.

I'm not able to repro this locally. Going to retry and see if CI hits it again.

**Comment by AndyAyersMS** [comment]
Stress failures so far are the known issues being disabled in #46350.

**Comment by BruceForstall** [comment]
Thanks for fixing this long-standing issue!

**Pull request merged**